File: btap-modeling/test/test_model_diagrams.rb
require_relative 'test_helper'

# The REUSABLE diagram API consumed by host reports (e.g. openstudio-necb's AHJ
# compliance report): BtapModeling.model_hvac_diagrams(model) draws the same
# OpenStudio-App-style loop diagrams the catalog draws, for ANY model — plus the
# self-contained icon defs + CSS a host document needs. SDK-only (no CLI).
class TestModelDiagrams < Minitest::Test
  include FixtureHelper

  # Build a hydronic boiler system on the fixture so there is a real hot-water
  # loop (pump -> boiler supply, baseboard demand) to diagram.
  def built_model
    model = load_fixture
    model.getThermalZones.each do |z|
      next if z.thermostatSetpointDualSetpoint.is_initialized

      z.setThermostatSetpointDualSetpoint(OpenStudio::Model::ThermostatSetpointDualSetpoint.new(model))
    end
    zones = model.getThermalZones.sort_by(&:nameString)
    BtapModeling.build_system(model, 'Baseboard gas boiler', zones)
    model
  end

  def test_model_hvac_diagrams_returns_loops_with_svg
    bundle = BtapModeling.model_hvac_diagrams(built_model)
    refute bundle[:empty], 'a boiler system has central loops'
    assert_operator bundle[:loops].size, :>=, 1, 'at least one loop diagrammed'
    loop = bundle[:loops].first
    assert loop[:kind], 'loop carries its kind'
    assert_kind_of String, loop[:label]
    assert_includes loop[:svg], '<svg', 'loop diagram is inline SVG'
    # the hot-water loop demand surfaces the served hydronic baseboards
    all_svg = bundle[:loops].map { |l| l[:svg] }.join
    assert_includes all_svg, 'Baseboard (hydronic)', 'demand branch lists served baseboards'
    # a boiler cell references the embedded boiler icon (resolves against icon_defs)
    assert_includes all_svg, '<use href="#icon-', 'cells reference embedded OS App icons'
  end

  # A single-zone (PSZ) air handler's loop label names the zone it serves, so a
  # host dropdown chooser can tell packaged single-zone units apart instead of
  # listing several ambiguous "Air loop" entries.
  def test_air_loop_label_names_its_served_zone
    model = load_fixture
    model.getThermalZones.each do |z|
      next if z.thermostatSetpointDualSetpoint.is_initialized

      z.setThermostatSetpointDualSetpoint(OpenStudio::Model::ThermostatSetpointDualSetpoint.new(model))
    end
    zone = model.getThermalZones.sort_by(&:nameString).first
    BtapModeling.build_system(model, 'PSZ RTU Gas and DX Coils and Hot Water Baseboard', [zone])
    bundle = BtapModeling.model_hvac_diagrams(model)
    air = bundle[:loops].find { |l| l[:kind] == :air }
    refute_nil air, 'a PSZ builds an air loop'
    assert_match(/\AAir loop — /, air[:label],
                 "single-zone air loop is labelled with its served zone, got #{air[:label].inspect}")
    assert_includes air[:label], zone.nameString, 'the label carries the served zone name'
  end

  def test_model_hvac_diagrams_never_raises_on_empty_model
    bundle = BtapModeling.model_hvac_diagrams(OpenStudio::Model::Model.new)
    assert bundle[:empty], 'a bare model has no HVAC'
    assert_empty bundle[:loops]
    assert_nil bundle[:zone_equipment_svg]
  end

  def test_icon_defs_are_embedded_once_and_self_contained
    defs = BtapModeling.hvac_icon_defs
    assert_includes defs, '<symbol id="icon-', 'icon symbols defined'
    assert_includes defs, 'data:image/png;base64,', 'icons embedded as data-URIs'
    # self-contained: no url(), no <link>/@import, no remote src/href
    refute_match(/@import|<link/, defs)
    refute_match(/url\(/, defs)
    refute_match(%r{(src|href)\s*=\s*"https?://}, defs)
  end

  def test_diagram_css_is_self_contained_and_sizes_svgs
    css = BtapModeling::CatalogReport::DIAGRAM_CSS
    assert_match(/\.diagram\s*\{[^}]*overflow-x:\s*auto/, css, 'scroll container')
    assert_match(/\.diagram\s+svg\s*\{[^}]*width:\s*auto/, css, 'intrinsic-size override')
    assert_match(/break-inside:\s*avoid/, css, 'print-friendly')
    refute_match(/url\(|@import/, css, 'no external references')
  end
end

File: python/tests/modeling/test_model_diagrams.py
"""The REUSABLE diagram API consumed by host reports (e.g. btap-necb's AHJ
compliance report): modeling.model_hvac_diagrams(model) draws the same
OpenStudio-App-style loop diagrams the catalog draws, for ANY model — plus the
self-contained icon defs + CSS a host document needs. SDK-only (no CLI).
"""

import unittest

import btap.modeling as modeling
from tests.support import load_fixture, needs_sdk


@needs_sdk
class TestModelDiagrams(unittest.TestCase):
    def built_model(self):
        """Build a hydronic boiler system on the fixture so there is a real hot-water
        loop (pump -> boiler supply, baseboard demand) to diagram."""
        import openstudio
        from btap._compat import sorted_by_name

        model = load_fixture()
        for z in model.getThermalZones():
            if z.thermostatSetpointDualSetpoint().is_initialized():
                continue

            z.setThermostatSetpointDualSetpoint(
                openstudio.model.ThermostatSetpointDualSetpoint(model))
        zones = sorted_by_name(model.getThermalZones())
        modeling.build_system(model, 'Baseboard gas boiler', zones)
        return model

    def test_model_hvac_diagrams_returns_loops_with_svg(self):
        bundle = modeling.model_hvac_diagrams(self.built_model())
        self.assertFalse(bundle['empty'], 'a boiler system has central loops')
        self.assertGreaterEqual(len(bundle['loops']), 1, 'at least one loop diagrammed')
        loop = bundle['loops'][0]
        self.assertTrue(loop['kind'], 'loop carries its kind')
        self.assertIsInstance(loop['label'], str)
        self.assertIn('<svg', loop['svg'], 'loop diagram is inline SVG')
        # the hot-water loop demand surfaces the served hydronic baseboards
        all_svg = ''.join(l['svg'] for l in bundle['loops'])
        self.assertIn('Baseboard (hydronic)', all_svg, 'demand branch lists served baseboards')
        # a boiler cell references the embedded boiler icon (resolves against icon_defs)
        self.assertIn('<use href="#icon-', all_svg, 'cells reference embedded OS App icons')

    def test_air_loop_label_names_its_served_zone(self):
        """A single-zone (PSZ) air handler's loop label names the zone it serves, so a
        host dropdown chooser can tell packaged single-zone units apart instead of
        listing several ambiguous "Air loop" entries."""
        import openstudio
        from btap._compat import sorted_by_name

        model = load_fixture()
        for z in model.getThermalZones():
            if z.thermostatSetpointDualSetpoint().is_initialized():
                continue

            z.setThermostatSetpointDualSetpoint(
                openstudio.model.ThermostatSetpointDualSetpoint(model))
        zone = sorted_by_name(model.getThermalZones())[0]
        modeling.build_system(model, 'PSZ RTU Gas and DX Coils and Hot Water Baseboard', [zone])
        bundle = modeling.model_hvac_diagrams(model)
        air = next((l for l in bundle['loops'] if l['kind'] == 'air'), None)
        self.assertIsNotNone(air, 'a PSZ builds an air loop')
        self.assertRegex(air['label'], r'\AAir loop — ',
                         "single-zone air loop is labelled with its served zone, "
                         f"got {air['label']!r}")
        self.assertIn(zone.nameString(), air['label'],
                      'the label carries the served zone name')

    def test_model_hvac_diagrams_never_raises_on_empty_model(self):
        import openstudio

        bundle = modeling.model_hvac_diagrams(openstudio.model.Model())
        self.assertTrue(bundle['empty'], 'a bare model has no HVAC')
        self.assertEqual([], bundle['loops'])
        self.assertIsNone(bundle['zone_equipment_svg'])

    def test_icon_defs_are_embedded_once_and_self_contained(self):
        defs = modeling.hvac_icon_defs()
        self.assertIn('<symbol id="icon-', defs, 'icon symbols defined')
        self.assertIn('data:image/png;base64,', defs, 'icons embedded as data-URIs')
        # self-contained: no url(), no <link>/@import, no remote src/href
        self.assertNotRegex(defs, r'@import|<link')
        self.assertNotRegex(defs, r'url\(')
        self.assertNotRegex(defs, r'(src|href)\s*=\s*"https?://')

    def test_diagram_css_is_self_contained_and_sizes_svgs(self):
        from btap.modeling.hvac.catalog_report import DIAGRAM_CSS

        css = DIAGRAM_CSS
        self.assertRegex(css, r'\.diagram\s*\{[^}]*overflow-x:\s*auto', 'scroll container')
        self.assertRegex(css, r'\.diagram\s+svg\s*\{[^}]*width:\s*auto',
                         'intrinsic-size override')
        self.assertRegex(css, r'break-inside:\s*avoid', 'print-friendly')
        self.assertNotRegex(css, r'url\(|@import', 'no external references')


if __name__ == '__main__':
    unittest.main()
